### ISSUE TITLE
Hang after running "history --help"

### ISSUE THREAD

**Issue by Russell-Jones** [xonsh/xonsh#1827]
This started happening on OS X 10.12, Ubuntu 14.04 and 16.04 after pulling from HEAD today. ^C stops it.

**Comment by gforsyth** [comment]
Thanks for reporting, @Russell-Jones!  I can confirm this.  And it looks like the same is true for `which --help`

**Comment by Russell-Jones** [comment]
I just had a go at tracking it down with git bisect, but too many versions didn't have the command working at all. I skipped those, and the result was 
````
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
83ecfd057855504874eb363142654f13f63d6aaf
1a34c10a39852c0d9015ff31b21e61941af69616
6180dbc70a4f3c5fd10eb4615e4a6cfb69598853
0e903dce8fdd4e7a6b3169e412d86311a4d81a54
2b5a5e2a750c63a78613e690abcf20ffade33118
0f80b67287b0633218879573307b474edd31e63c
d135f17cf3503466f0c0d8f5fb193d01bd39fdbd
5cd40eb13ff02eff98c567d3aed3967cd4765539
d415a867a9090df187e67cbaf366ef4a9dba1d91
643615ab983c0164b5605e4b11124ba0334b309c
911021e0c3908803e063f841c8dfd119613777e3
2ebd81e2b008fc098dcdf0f432ecbfa351949d34
c5288df4b5a9ca35e614861009e4aa0c77922680
4e012be3b4ee6303d36c29738c1d5dd58e89d2eb
b29c456f4d4323c6fe8f1ae4640ad95cbabca354
5ee1d81cbc7c23731099f0b1fbd24171126892ff
f8cb5392c49126e0c2ac31cefdf9bb2154c98954
2b3ae98289d422db83572c4c5e39a45f55dfef58
51866ee92e99415a49259824642c10d3c8673b4c
bde2f985e69c8d0fb694fbd9f3dd3f1d2307a0ab
c3ec3dd1e60e4cf260f9dc91608e9874ae65cdd1
65113eb7935cc78f4b9a827cce46a5d53df6bbe1
861aedfc4dc2d3e3630db27b625c7ae1c69595d6
9e4dbd446ee82f6a2615e15029285087d7b8a15d
d999c4fb2664c707abed6ab65761d609573b6824
27216528c3df678374b273604c5829d28bc2bf06
a59aa59e84c470da78b7f814f8cfcf7fc26c3d8d
24e0f6180879cbea5277bb87d13a2efbf721bd09
ccaca296625db071c3911bb1c07a6c9c2c7d07c8
5a7e2b2784e32a70134d8dbb6d6e14da569db270
e2c0c29d8d70f8fa893f7f56de2a02341bb0369d
6f968fc80c379fa2fe227eafc5400135831ecc3c
afccd10b6a4f1868d458b521a293070f65321641
2f2f8db47137df50519dbd8c66b5ad0f83edb231
a18950c126df874f38e15f483cae56de56010542
a6de9f3841bedc8dfd5dadf883d5ea50a8a120db
c5d014b730aabfd2729b2b77074cf1af1515153f
e50c99806d7291c4cb4278f1e0dc446131a52619
fb9d55c343e760e88603d1c8b131f38062f3587b
78880fadd9fa724c59e994697787418dcd1fe017
cc7aa92db23e2bb66978711000f4918be4a46d7b
44510bf0ce943ad5ed03a98a8c511c474753f867
5037a60f044bf155b501d0e7a40a604e39ca3950
0b8505ec176a108825591891b388b37ca7cc511a
dbb349831ba0f7d456f0c2334a7fdeebe86af980
04e781ffdcf84a12e8290e1c8f31870cf891e9bb
20a1c01bc55bb30e1fc423880f5879813186ab85
91447afa595655fe24ad06b3d2b316e986b8c2cd
f1c9765492cd137d9fa08030ae45f9d21032e5a7
7dc0e95fb6b1fe6b19a724929f0bc44197225681
85d3d3d65ae1d158e9f2b612edc7106094ac74a4
05954f480d7113d22799777ed6c3f99cce5077f9
a4bceb4e0b62c8c298e53441ab590a043e930341
2cddac0afb420d1aae0136677cf1fc688b59be72
b0951ec82f4873e42899876b8d865728944485a5
69048980333dd213ba0ed88b21fcf93ea25a3430
8dfd594104342df832fccda48a0bbdafac42e0e2
50e6faee8705158bebe317173be86df3e9584f09
We cannot bisect more!
````
Hope that helps :)

**Comment by blahgeek** [comment]
This is directly related to #1819 . After #1758 , `sys.exit()` is not properly handled on function alias. While running `history --help` or `which --help` or any function alias that calls `ArgumentParser`, `sys.exit` is called inside `parse_args` and causes this issue.

**Comment by scopatz** [comment]
Ahh, I think that you are right about sys.exit() taking out the whole process.  This is a general no-no and a huge mistake for argparser to be doing.  That said, we do need to be robust to it.

**Comment by scopatz** [comment]
And looking at this, I am not sure how it ever worked...

**Comment by scopatz** [comment]
Alright, a potential fix is in #1841 for the sys.exit() stuff.

**Comment by gforsyth** [comment]
Closed via #1841
